### ISSUE TITLE
Consuming ARM template outputs crash

### ISSUE THREAD

**Issue by Rurouni** [hashicorp/terraform#7353]
Hi,
I am trying to use terraform for azure arm, but as soon as I used output is has crashed... I suppose it's because I am doing something weird with mapping outputs  of ARM template but there is not much documentation on the topic..
### Terraform Version

Terraform v0.6.16
### Affected Resource(s)

Please list the resources as a list, for example:
- azurerm_template_deployment
### Terraform Configuration Files

``` hcl
provider "azurerm" {
  <<hidden>>
}

variable "location" {
    default = "<<hidden>>"
}

variable "groupName" {
    default = "<<hidden>>"
}

# Create a resource group
resource "azurerm_resource_group" "monitoring" {
    name        = "${var.groupName}"
    location    = "${var.location}"
}

resource "azurerm_template_deployment" "monitoring" {
    name = "monitoring"
    resource_group_name = "${azurerm_resource_group.monitoring.name}"
    template_body = <<DEPLOY
{
  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "storageAccountType": {
      "type": "string",
      "defaultValue": "Standard_LRS",
      "allowedValues": [
        "Standard_LRS",
        "Standard_GRS",
        "Standard_ZRS"
      ],
      "metadata": {
        "description": "Storage Account type"
      }
    }
  },
  "variables": {
    "location": "[resourceGroup().location]",
    "storageAccountName": "[concat(uniquestring(resourceGroup().id), 'storage')]",
    "publicIPAddressName": "[concat('myPublicIp', uniquestring(resourceGroup().id))]",
    "publicIPAddressType": "Dynamic",
    "apiVersion": "2015-06-15",
    "dnsLabelPrefix": "terraform-acctest"
  },
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "name": "[variables('storageAccountName')]",
      "apiVersion": "[variables('apiVersion')]",
      "location": "[variables('location')]",
      "properties": {
        "accountType": "[parameters('storageAccountType')]"
      }
    },
    {
      "type": "Microsoft.Network/publicIPAddresses",
      "apiVersion": "[variables('apiVersion')]",
      "name": "[variables('publicIPAddressName')]",
      "location": "[variables('location')]",
      "properties": {
        "publicIPAllocationMethod": "[variables('publicIPAddressType')]",
        "dnsSettings": {
          "domainNameLabel": "[variables('dnsLabelPrefix')]"
        }
      }
    }
  ],
  "outputs":{
    "publicUrl":{
        "type":"string",
        "value" : "[reference(variables('publicIPAddressName')).dnsSettings.fqdn]"
      },
    "testObject":{
        "type":"object",
        "value" : "[reference(variables('storageAccountName'))]"
      }
  }
}
DEPLOY
    deployment_mode = "Complete"
}

output "azure.RG.monitoring" {
    value = "${azurerm_template_deployment.monitoring.outputs.publicUrl}"
}
```
### Panic Output

https://gist.github.com/Rurouni/c1e997942576ae68b8e112952326a768
### Steps to Reproduce
1. `terraform apply`

**Comment by stack72** [comment]
Hi @Rurouni 

```
"testObject":{
        "type":"object",
        "value" : "[reference(variables('storageAccountName'))]"
      }
```

Why is the output an object here? Are you trying to dump the entire storage_account tree? Is this supported by ARM templates (outside of Terraform)

Thanks

Paul

**Comment by Rurouni** [comment]
Hi Paul,
Yes it is supported, if you have vaule of "object" in ARM script then output will contain json object with all the properties that were used in that script(sadly not whole tree as it would make discovering potential ouptuts from ARM so much easier)

**Comment by alangber** [comment]
Hi Paul,
Is there any progress on this issue? I'm getting the same error and would really like this to work...

**Comment by bpoland** [comment]
Not sure if I'm doing something differently but this is working for me with terraform 0.9.1 -- I specify the output in the ARM template, and am able to reference it in terraform outputs or other resources.

**Comment by bpoland** [comment]
Actually seems like it only works if you add the output after the template has already been deployed (then it will have the output available).  If you do it before the template has been deployed, you get this:

Error running plan: 1 error(s) occurred:

* dns_a_record_set.appgw: 1 error(s) occurred:

* dns_a_record_set.appgw: Resource 'azurerm_template_deployment.vmss_appgw' does not have attribute 'outputs.appGwIp' for variable 'azurerm_template_deployment.vmss_appgw.outputs.appGwIp'

**Comment by StephenWeatherford** [comment]
@Rurouni and @alangber  The only issues I see with outputs is if they are of type object, can you confirm that's what you are seeing?  Thx.

@bpoland Would you mind sharing your template file or at least the outputs portion that's causing the problem (minus secrets)?  Thx.

Stephen

**Comment by bpoland** [comment]
Output section of my ARM template is below.  The issue seems to be that before terraform actually deploys the template, the outputs.appGwIp object doesn't exist so terraform complains.  Once terraform has deployed the template, the outputs.appGwIp object exists and it's fine.

```
"outputs": {
        "appGwIp": {
            "type": "string",
            "value": "[reference(variables('appGwName')).frontendIpConfigurations[0].properties.privateIpAddress]"
        }
    }
```

**Comment by Rurouni** [comment]
@StephenWeatherford I am not using terraform with azure for quite a while, so don't know if the bug is still there, but originally it was crashing only on "object" output type, scalar types were working fine.

**Comment by Rurouni** [comment]
@StephenWeatherford I am not using terraform with azure for quite a while, so don't know if the bug is still there, but originally it was crashing only on "object" output type, scalar types were working fine.

**Comment by StephenWeatherford** [comment]
Moved bpoland's issue of getting an error before deployment to a separate issue: #13437

**Comment by StephenWeatherford** [comment]
@Rurouni Could you go ahead and close this, please?  The crash has been fixed, and int and bool support has been added.  Thx.

**Comment by stack72** [comment]
Closed via #13670
